src/aws-route53resolver/resolver-endpoint.ts

```diff
@@ -0,0 +1,389 @@
+   */
+  readonly vpc: IVpc;
+  /**
+   * The name of this Resolver Endpoint.
+   *
+   * @default - Generate a new name bu AWS CDK.
+   */
+  readonly resolverEndpointName?: string;
+  /**
+   * One or more security groups that control access to this VPC.
+   * The security groups rules must allow TCP and UDP access.
+   *
+   * **For Inbound Endpoint**
+   * - The security group must include one or more inbound rules.
+   * - Open port 53.
+   *
+   * **For Outbound Endpoint**
+   * - The security group must include one or more outbound rules.
+   * - Open the port that you're using for DNS queries on your network.
+   *
+   * @default - Create a new Security Group.
+   */
+  readonly securityGroups?: ISecurityGroup[];
+  /**
+   * The subnets and IP addresses. We require you to specify a minimum of two IP addresses for redundancy.
+   *
+   * **For Inbound Endpoint**
+   * - The IP addresses that you want DNS resolvers on your network to forward DNS queries to.
+   *
+   * **For Outbound Endpoint**
+   * - The IP addresses in your VPC that you want Resolver to forward DNS queries to on the way to resolvers on your network.
+   * These are not the IP addresses of the DNS resolvers on your network, you specify resolver IP addresses when you create the rules that you associate with one or more VPCs.
+   *
+   * @default - Auto assign from private subnets.
+   */
+  readonly ipAddresses?: IpAddresses;
+  /**
+   * For the IP address type you can choose either IPv4, IPv6, or dual-stack.
+   * A dual-stack endpoint means that it will resolve via both IPv4 and IPv6.
+   * This IP address type is applied to all IP addresses.
+   *
+   * @default IpAddressType.IPV4
+   */
+  readonly ipAddressType?: IpAddressType;
+}
+
+/**
+ * Resolver Endpoint direction.
+ */
+enum Direction {
+  /**
+   * Direction of inbound.
+   */
+  INBOUND = 'INBOUND',
+  /**
+   * Direction of outbound.
+   */
+  OUTBOUND = 'OUTBOUND',
+}
+
+/**
+ * Properties for a Resolver Endpoint.
+ */
+interface ResolverEndpointBaseProps extends ResolverEndpointOptions {
+  readonly direction: Direction;
+  readonly protocols?: Protocols;
+  /**
+   * @default false
+   */
+  readonly allowAllOutbound?: boolean;
+  /**
+   * @default false
+   */
+  readonly allowAllIpv6Outbound?: boolean;
+}
+
+abstract class ResolverEndpointBase
+  extends Resource
+  implements IConnectable, IResolverEndpoint {
+  readonly connections: Connections;
+  readonly resolverEndpointArn: string;
+  readonly resolverEndpointId: string;
+  constructor(scope: Construct, id: string, props: ResolverEndpointBaseProps) {
+    super(scope, id, {
+      physicalName: props.resolverEndpointName,
+    });
+    this.connections = new MultiDefaultPortConnections({
+      // Will be replace to Port.DNS_TCP and Port.DNS_UDP
+      defaultPorts: [Port.tcp(53), Port.udp(53)],
+      securityGroups: props.securityGroups ?? [
+        new SecurityGroup(this, 'SecurityGroup', {
```

Comment:
Please extract the creation of sub-constructs into a factory method that can be overridden in a subclass